File: data/prepare_data.js
const XLSX = require("xlsx");
const fs = require("fs");

const excelPath = "../core_career_skill_dataset_39.xlsx";

const workbook = XLSX.readFile(excelPath);

const sheet = workbook.Sheets["Core Skills Dataset"];

const data = XLSX.utils.sheet_to_json(sheet);

console.log(`Found ${data.length} skill records.`);

// -----------------------------
// Careers
// -----------------------------

const careerNames = [
    ...new Set(data.map(row => row.Career))
].sort();

const careers = careerNames.map((name, index) => ({
    id: index + 1,
    name
}));


// -----------------------------
// Skills
// -----------------------------

const skillMap = new Map();

data.forEach(row => {
    const key = `${row["Core Skill"]}|${row.Category}`;

    if (!skillMap.has(key)) {
        skillMap.set(key, {
            name: row["Core Skill"],
            category: row.Category
        });
    }
});

const skills = [...skillMap.values()]
    .sort((a, b) => a.name.localeCompare(b.name))
    .map((skill, index) => ({
        id: index + 1,
        ...skill
    }));


// -----------------------------
// Career-Skills
// -----------------------------

const careerSkills = data.map((row, index) => {

    const career = careers.find(
        c => c.name === row.Career
    );

    const skill = skills.find(
        s =>
            s.name === row["Core Skill"] &&
            s.category === row.Category
    );

    return {
        id: index + 1,
        career_id: career.id,
        skill_id: skill.id,
        required_level: row["Required Level"],
        priority: row.Priority,
        suggested_project: row["Suggested Project"]
    };
});


// -----------------------------
// CSV converter
// -----------------------------

function toCSV(rows) {

    if (rows.length === 0) return "";

    const headers = Object.keys(rows[0]);

    const escape = value => {
        if (value === null || value === undefined) {
            return "";
        }

        const text = String(value);

        if (
            text.includes(",") ||
            text.includes('"') ||
            text.includes("\n")
        ) {
            return `"${text.replace(/"/g, '""')}"`;
        }

        return text;
    };

    const lines = [
        headers.join(","),
        ...rows.map(row =>
            headers
                .map(header => escape(row[header]))
                .join(",")
        )
    ];

    return lines.join("\n");
}


// -----------------------------
// Save files
// -----------------------------

fs.writeFileSync(
    "careers.csv",
    toCSV(careers)
);

fs.writeFileSync(
    "skills.csv",
    toCSV(skills)
);

fs.writeFileSync(
    "career_skills.csv",
    toCSV(careerSkills)
);


console.log("\nData preparation complete!");
console.log(`Careers: ${careers.length}`);
console.log(`Skills: ${skills.length}`);
console.log(`Career-skill relationships: ${careerSkills.length}`);

console.log("\nCreated:");
console.log("✓ careers.csv");
console.log("✓ skills.csv");
console.log("✓ career_skills.csv");

File: data/prepare_data.py
import pandas as pd

excel_path = "../core_career_skill_dataset_39.xlsx"

# Read the main sheet
df = pd.read_excel(
    excel_path,
    sheet_name="Core Skills Dataset"
)

# -------------------------------------------------
# 1. Careers
# -------------------------------------------------

careers = (
    df[["Career"]]
    .drop_duplicates()
    .sort_values("Career")
    .reset_index(drop=True)
)

careers.insert(0, "id", range(1, len(careers) + 1))

careers.to_csv("careers.csv", index=False)


# -------------------------------------------------
# 2. Skills
# -------------------------------------------------

skills = (
    df[["Core Skill", "Category"]]
    .drop_duplicates()
    .sort_values("Core Skill")
    .reset_index(drop=True)
)

skills.insert(0, "id", range(1, len(skills) + 1))

skills = skills.rename(columns={
    "Core Skill": "name",
    "Category": "category"
})

skills.to_csv("skills.csv", index=False)


# -------------------------------------------------
# 3. Career-Skills relationship
# -------------------------------------------------

career_skills = df.merge(
    careers,
    on="Career"
)

career_skills = career_skills.merge(
    skills,
    left_on=["Core Skill", "Category"],
    right_on=["name", "category"]
)

career_skills = career_skills.rename(columns={
    "id_x": "career_skill_id",
    "id_y": "skill_id",
    "Required Level": "required_level",
    "Priority": "priority",
    "Suggested Project": "suggested_project"
})

career_skills = career_skills[
    [
        "career_skill_id",
        "id",
        "skill_id",
        "required_level",
        "priority",
        "suggested_project"
    ]
]

career_skills = career_skills.rename(columns={
    "id": "career_id"
})

career_skills.to_csv(
    "career_skills.csv",
    index=False
)


# -------------------------------------------------
# Done
# -------------------------------------------------

print("Data preparation complete!")
print()
print(f"Careers: {len(careers)}")
print(f"Skills: {len(skills)}")
print(f"Career-skill relationships: {len(career_skills)}")
print()
print("Created:")
print("- careers.csv")
print("- skills.csv")
print("- career_skills.csv")